models/icepack/model_mod.f90

```diff
@@ -0,0 +1,885 @@
+
+allocate(TLAT(Nx), TLON(Nx))
+
+call read_horiz_grid(Nx, TLAT, TLON)
+ 
+if (do_output()) write(logfileunit, *) 'Using grid : Nx, Ncat = ', Nx, Ncat
+if (do_output()) write(*, *) 'Using grid : Nx, Ncat = ', Nx, Ncat
+
+domain_id = add_domain('cice.r.nc', nfields, &
+                       var_names = variable_table(1:nfields, VAR_NAME_INDEX), &
+                       kind_list = state_kinds_list(1:nfields), &
+                       update_list = update_var_list(1:nfields))
+
+if (debug > 2) call state_structure_info(domain_id)
+
+model_size = get_domain_size(domain_id)
+if (do_output()) write(*, *) 'model_size = ', model_size
+
+
+end subroutine static_init_model
+
+!------------------------------------------------------------------
+! Returns the number of items in the state vector as an integer. 
+
+function get_model_size()
+
+integer(i8) :: get_model_size
+
+get_model_size = model_size
+
+end function get_model_size
+
+!------------------------------------------------------------------
+! Given a state handle, a location, and a model state variable type,
+! interpolates the state variable fields to that location and returns
+! the values in expected_obs. The istatus variables should be
+! returned as 0 unless there is some problem in computing the 
+! interpolation, in which case an alternate value should be returned.
+
+subroutine model_interpolate(state_handle, ens_size, location, obs_type, expected_obs, istatus, thick_flag)
+
+type(ensemble_type), intent(in) :: state_handle
+integer,             intent(in) :: ens_size
+type(location_type), intent(in) :: location
+integer,             intent(in) :: obs_type
+real(r8),           intent(out) :: expected_obs(ens_size) ! array of interpolated values
+integer,            intent(out) :: istatus(ens_size)
+logical,optional,    intent(inout) :: thick_flag
+
+!local vars
+real(r8)       :: loc_array(3), llon, llat
+integer(i8)    :: base_offset
+integer        :: cat_index, cat_signal, icat, cat_signal_interm
+real(r8)       :: expected_aggr_conc(ens_size)
+integer        :: set_obstype
+
+!Fei---need aicen*fyn to calculate the aggregate FY concentration
+real(r8)       :: expected_conc(ens_size)
+real(r8)       :: expected_fy(ens_size)
+real(r8)       :: expected_tsfc(ens_size)
+real(r8)       :: temp(ens_size)
+real(r8)       :: temp1(ens_size)
+
+if ( .not. module_initialized ) call static_init_model
+
+expected_obs(:) = MISSING_R8 ! represents a bad value in DART
+istatus(:) = 99
+
+loc_array = get_location(location)
+llon    = loc_array(1)
+llat    = loc_array(2)
+cat_index = int(loc_array(3))
+
+if (obs_type == QTY_SEAICE_CATEGORY) then
+   if (cat_index <= Ncat) then
+      istatus      = 0
+      expected_obs = cat_index
+      RETURN
+   endif
+endif
+if (debug > 1) then
+   print *, 'requesting interpolation of ', obs_type, ' at ', llon, llat, cat_index
+endif
+
+SELECT CASE (obs_type)
+   CASE (QTY_SEAICE_AGREG_THICKNESS )  ! these kinds require aggregating 3D vars to make a 2D var
```

Comment:
It would be good to write out the psuedo code for what this interpolation is doing.
so a description of the state vector, and for for given QTY how is the model_interpolate calculated.